bubblewrap.c

```diff
@@ -1628,6 +1691,32 @@ warn_only_last_option (const char *name)
   warn ("Only the last %s option will take effect", name);
 }
 
+static void
+make_setup_overlay_src_ops (const char *const *const argv)
+{
+  /* SETUP_OVERLAY_SRC is unlike other SETUP_* ops in that it exists to hold
+   * data for SETUP_{,TMP_,RO_}OVERLAY_MOUNT ops, not to be its own operation.
+   * This lets us reuse existing code paths to handle resolving the realpaths
+   * of each source, as no other operations involve multiple sources the way
+   * the *_OVERLAY_MOUNT ops do.
+   *
+   * While the --overlay-src arguments are expected to precede the
+   * --overlay argument, in bottom-to-top order, the SETUP_OVERLAY_SRC ops
+   * follow their corresponding *_OVERLAY_MOUNT op, in top-to-bottom order
+   * (the order in which overlayfs will want them). They are handled specially
+   * in setup_new_root () during the processing of *_OVERLAY_MOUNT.
+   */
+  int i;
+  SetupOp *op;
+
+  for (i = 1; i <= next_overlay_src_count; i++)
+    {
+      op = setup_op_new (SETUP_OVERLAY_SRC);
+      op->source = argv[1 - 2 * i];
```

Comment:
Maybe explain why this is safe (all the `--overlay-src` args are always directly before the finishing arg).

I really do wonder if adding the ops right away and then processing them in `setup_newroot` would not be smarter.

Reply:
Does ‘While the --overlay-src arguments are expected to precede the --overlay argument’ in the immediately preceding comment not communicate that?

It's true I could have stuffed these ops in a list first and then reversed them in `setup_newroot`... but this depends only on libc, so first I would have had to invent a list. And then I would have had to write new code for resolving all the paths associated with the overlay op, because all the other ops are only associated with at most one host path. Reusing the op list to store the `overlay_src` data kills both birds with one stone and a minimum of new code. In a larger application with more moving parts, I might have done it the other way, but I don't think it's too difficult to keep track of what's going on here (and the big comment is there to warn readers to watch their heads).

Reply:
Preceding defines the order but not if there are other arguments in between.

There already is a linked list that could be used but I see your point. No strong objection from me.

Reply:
> There already is a linked list that could be used

Ah, you're right about that! My mistake; in my defense, it's been over half a year since I did anything non-trivial with this code.